### DISCUSSION TITLE
Developers conversations

### DISCUSSION THREAD

**Discussion by leikoilja** [leikoilja/ha-google-home#23] [General]
## 👋 Welcome!

Since we are multiple developers from all over the place let's try using github's "Discussions" feature to chat about any issues, ideas or things we want to, before we take it to the "Projects" board and create reals tasks (aka issues) from them.

I have never used this feature before, but it would be great to try it as a tool while collaborating 😄

**Comment by leikoilja** [comment]
@DurgNomis-drol, sorry, I ll move your "off-topic" question from the issue here, to the discussion.

> You were basically asking if we are aiming to release it as a custom component or to merge it into the core of Home Assistant.

@DurgNomis-drol, it's up to us to decide, but apart of convenience, I see no point of trying to get it into HA core at this point. I am currently working on refactoring `habitica` integration in HA `core` and my PR was work in progress ever since August 2020. It will take enormous effort and much more refactoring to implement it in the `core` repo. I propose we just focus on making it in sort-of following stages:
- Making it work together with `glocaltokens`
- Finding a way to eliminate the need to manually installing system-wise compliler libraries (like `gcc`, `g++`, `linux-headers`) - maybe by opening issues at HA repos to propose to include these libraries as defaults when building containers
- Release this integration into `HACS` store
- Get some users testing it and refactor/fix/enhance when more usage is seen
- And only then get back to this discussion to decide together if we want to move it into HA `core`

**Reply by DurgNomis-drol** [reply]
This sounds like a wise choice for now. :smile: I wasn't think of doing it straight away anyway, so lets leave it as you say for now.


This is an awesome feature that I didn't know existed! :smile:

**Comment by ArnyminerZ** [comment]
About the HA `zeroconf` implementation: I think that should be the way. Right now we just needed a way of getting the IP address of the device, which has been added in `glocaltokens`, so we can keep working on `ha-glocaltokens`. I think once we have the base of `ha-localtokens` working well in the required devices, including the RPis, which wasn't working, and all the main features implemented, we can start working on the discovery, since it's more like an improvement, something to make the integration easier to configure, than an indispensable thing.

**Reply by DurgNomis-drol** [reply]
I think I got things confused in the other thread.
Do you mean auto discovery of the integration (as in checks the local network if a google home device exists and the asks the user to setup the integration)? 😄

**Reply by ArnyminerZ** [reply]
Yes, that's what's referred to in https://developers.home-assistant.io/docs/creating_integration_manifest/#zeroconf

**Reply by DurgNomis-drol** [reply]
Okay, i got confused and i agree with you statement above. I'm a newbie when it comes both python and HA development, but im learning a lot 😁

**Reply by ArnyminerZ** [reply]
Yeah, I've used python for quite a long time, but it keeps confusing me in some things. It's really different from good old languages such as Java, or all the Cs

**Reply by DurgNomis-drol** [reply]
I'm an electrian in my daily work, so this is only for fun. I have never had any formal training in programing, so sometimes i have read up on some of the things you guys talk about 😉

I have a hardtime understanding C and Java.

**Reply by ArnyminerZ** [reply]
Don't worry, it's normal to get confused, I said the Cs and Java because they are older, and people usually like them just for being older. Just joking 😆

**Reply by leikoilja** [reply]
haha, as long as it's not some sort of sketchy css, I'm game 😄  i always get confused with design part, especially when they start putting in some hard complex structures, css-processors etc etc 😄

**Reply by ArnyminerZ** [reply]
I happens to me the same, I can never manage to create beautiful UIs 😢 I always end up using classics like Bootstrap

**Comment by leikoilja** [comment]
Boooom, @DurgNomis-drol, @ArnyminerZ! 
Have just merged in `dev` branch making it pre-release of v2 of our this package! Let's work on `master` from now on with proper PRs and issues 🚀 
Needed to do that to finalize the renaming of the project, now we are under - https://github.com/leikoilja/ha-google-home 💥 

Please make sure the fork you have is working well and the root directory is renamed to `ha-google-home`

**Comment by leikoilja** [comment]
Also, an interesting resource I came across is integration score - https://developers.home-assistant.io/docs/integration_quality_scale_index/ 💥

**Reply by DurgNomis-drol** [reply]
Good find! :smile: 

I think it a good idea for us to have this in minded when further developing the integration, so if we eventually decides to try to get it included in HA, it will be a lot easier.

**Comment by leikoilja** [comment]
Okaayz, I'm back home to my setup and have some great news! Python wheels PR is successfully merged now, so we DO NOT have to install custom system-wide libraries(`gcc`, `g++`, `linux-headers`) no more 💥  And all the actions pass successfully on this repo, wohoo! 🥳 

All we need to do now to install HA-Google-Home on Homeassitent is to simply add a custom repository(https://github.com/leikoilja/ha-google-home) into `HACS` store. When you enter HACS store, in the top right corner under `...` dots you can add custom repository. Once custom repository is added, restart HA and simply install`HA-Google-Home` by adding new Integration menu. 

Not that that is over, we can finally focus on refactoring, adding features, and finally releasing the first BETA versions for testers, who are keen on trying it out! 💯 

Once we made a first beta release, I'll continue working to add `HA-Google-Home` into the default HACS store.

Please note, `master` branch is currently broken and I'm doing my best to fix it and make it work with a simplest feature set to begin with. Once fixed, we will be working on `dev` branch to keep `master` always stable and introduce changes incrementally as new releases.

Thanks for the awesome work, @ArnyminerZ, @DurgNomis-drol 💯  🥳  Let's keep it going!

**Comment by leikoilja** [comment]
Awesome, we are on a very good track! Now that the PR is merged, the `master` should be up and running! 💥 
@DurgNomis-drol can you please help testing it whenever you have time? Would also much appreciate it if you could help to update the `README` documentation to prepare it for the first beta release. We can probably almost rewrite the entire README file, adding instructions how to install using HACS (by adding a custom repo) and some info about timers/alarms and updating stuff about token usage for custom API calls. We can merge those documentation changes in and then release our first beta very soon 💯

**Reply by DurgNomis-drol** [reply]
Yes, i will test it to night. I run Homeassistant in a container so i can't help with the HACS part of the readme, but I also think we should provide a installation method for those who use it in docker (like me)

**Reply by leikoilja** [reply]
Ou, that's good to know. Yeh, let's keep both options available - HACS and manual installation :)

**Reply by DurgNomis-drol** [reply]
I can confirm that installing it manual seems too work, `grpcio`, `grpcio-tools` and` glocaltokens` installs automatically. 

I will make changes to README that describes how to do that :)

See #51 for an issue i found

**Reply by DurgNomis-drol** [reply]
Do you now what the minimum required version of HA is? I have only tested it with 2021.3 because I wouldn't install below that

**Reply by leikoilja** [reply]
No, I don't... but let's just put the current version that we have tested it with and support further on 🤷‍♂️😂

**Comment by leikoilja** [comment]
Wohoooo, our FIRST [release](https://github.com/leikoilja/ha-google-home/releases) is out! So so happy! 
Massive thanks to @KapJI, @DurgNomis-drol, @ArnyminerZ 💥 🎉

**Reply by KapJI** [reply]
Awesome! 🚀
Please also update `manifest.json` 🙂

**Reply by leikoilja** [reply]
Yep, thanks for the spot! Updated ✅

**Reply by DurgNomis-drol** [reply]
Awesome! 🎉 🥳

**Reply by DurgNomis-drol** [reply]
I just installed v1 and it just works 😄 This has been my first project and i have learned so much from it 🎉 

I think everybody did an awesome job!

**Comment by leikoilja** [comment]
Hey-hey dev team! 💥
Some super exciting news! I'm stocked to announce that a talented developer @Drakulix is also joining our team after [discussions](https://github.com/Drakulix/googlehome/issues/23) to merge two ongoing projects `ha-google-home` and [googlehome](https://github.com/Drakulix/googlehome) 🚀 
He was actively working on a fork of the original HA Google Home component before its depreciation introducing some near workarounds. 
Welcome, @Drakulix 💥 💯

**Reply by KapJI** [reply]
Welcome, @Drakulix!

**Reply by KapJI** [reply]
I like the idea that API calls should be moved to external package like `googledevices`, I think we eventually should do that.

**Reply by Drakulix** [reply]
Thanks for the kind introduction! Hey everybody :wave:

**Reply by DurgNomis-drol** [reply]
@Drakulix Welcome to the team 😁

**Reply by ArnyminerZ** [reply]
> I like the idea that API calls should be moved to external package like `googledevices`, I think we eventually should do that.

I also think it should be done, but isn't that what `glocaltokens` does? Or do we wanna keep `glocaltokens` as the "token getter", and create a new package like `googledevices` as you have stated for the API calls?
Maybe we should open an issue for discussing this, or even an exclusive discussion since I believe this is quite an important change.

**Comment by leikoilja** [comment]
@KapJI, @DurgNomis-drol, I m thinking we need a new release, what do you say? My HA setup is sadly still broken 😅

**Reply by DurgNomis-drol** [reply]
Did 2021.4 break your setup? 😞 

Yeah i think that would be great 😃

**Comment by ArnyminerZ** [comment]
@leikoilja @KapJI @DurgNomis-drol I'm starting to develop some scripts for loading Eureka into the Integration, but I don't think every single element from it should be stored, since as you can see in [GHKAPI](https://rithvikvibhu.github.io/GHLocalApi/#operation/EurekaInfo) there are quite a few, which ones do you think that we should load in?

I put here a list with all the available parameters, and I believe we should choose which ones to include and which not:
- [ ] `version`
- [ ] `audio`
- [ ] `name`
- [ ] `build_info`
- [ ] `detail`
- [ ] `device_info`
- [ ] `net`
- [ ] `wifi`
- [ ] `setup`
- [ ] `settings`
- [ ] `opt_in`
- [ ] `opencast`
- [ ] `multizone`
- [ ] `proxy`
- [ ] `night_mode_params`
- [ ] `user_eq`
- [ ] `room_equalizer`
- [ ] `sign`
- [ ] `aogh`
- [ ] `ultrasound`
- [ ] `mesh`

It might even be possible to include all of them, maybe on different entities, but I find it too much and unnecessary, what do you think?

**Reply by KapJI** [reply]
Why do you want to load them in the first place? 🙂

**Reply by ArnyminerZ** [reply]
@KapJI 
Issues #248 and #199, among others, have stated the issue around the risk of having multiple devices with the same sames, which as discussed in [glocaltokens#75](https://github.com/leikoilja/glocaltokens/issues/75) should be fetched from Eureka, so once we load one parameter, I think we should also load the "compatibility table", to check if the device is compatible with night mode, bluetooth, and others (#220).
So as far as I see it, we should create a new model for the Eureka data, well, at least not the full data, but just the one we need. This was discussed to be added in `glocaltokens`, but since `glocaltokens` "is not loading any data", I think we should keep it in this package, until the API is moved to a new sepparate package, which was discussed somewhere that I can't find, too much discussions around here 😅, but @leikoilja suggested loading some Eureka data in a [comment in #220](https://github.com/leikoilja/ha-google-home/issues/220#issuecomment-822372790).

**Reply by KapJI** [reply]
I think you can create a method which queries eureka info, stores the full dict in device class but don't expose it to the UI unless needed. Then you can use some values from it in the code.

**Reply by ArnyminerZ** [reply]
Yeah, but the Eureka Info is really large, like, a lot xDD, do we really want to create a dict that large?

**Reply by KapJI** [reply]
If you query it anyway, there's very little overhead to store it. It's Python, not C++, we don't need to save bytes :)

**Reply by KapJI** [reply]
For my Google Home Mini this is about 1.2 KB of data for basic query and 8 KB for all possible data. We can do that 🙂
But don't fetch sections which you don't need. I think just `device_info` should be enough.

**Reply by ArnyminerZ** [reply]
I'll be opening a new PR just for following development

**Reply by leikoilja** [reply]
You are right, we probs don't need all the data there is in the `eureka`, but it would be super nice to make the implementation scalable with a simple list of "values" that needs to be loaded. So when we need more data from `eureka` it would be as simple as adding a new record to the list. 

It would indeed be a good idea to store it to a separate package after all. 
Since we already have a full control and close coupling of `glocaltokens` perhaps we can discuss repurposing `glocaltokens` into something a bit more generic where we can offload more API related calls. What do you think?

**Reply by ArnyminerZ** [reply]
I like both the ideas of loading the data from glocaltokens and from another package. For me, glocaltokens is looking really clean, so adding more code like the API loading to glocaltokens seems a bit excessive for me, so I'd create a new package. Whatsoever, both options should work.

**Reply by leikoilja** [reply]
My concerns are that it become too many packages around doing similar things :D makes it harder to maintain when things are scattered. `glocaltokens` is called `tokens` but in reality we are already doing more then just tokens (we fetch google devices from homegraph, their corresponding tokens and device models, do zeroconf to find their local IP).

There is already a [googledevices](https://github.com/ludeeus/googledevices) package out there that does **_some_** API calls and seems like have some good coding practises in place. It's not maintained, so I would be more happy if we fork or "rewrite" our own implementation (either into `glocaltokens` and rename it to `...`(smth more generic)) or creating a new package that we'll use in `ha-google-home` later. That would leave a complete control of repo/project to us making it much easier to maintain and work with

**Reply by KapJI** [reply]
I think it's cleaner to have isolated scope for each package. E.g. glocaltokens should only interact with HomeGraph API to fetch local tokens. Zeroconf is probably better to be moved to component itself and pass discovered devices to glocaltokens. API calls to actual devices can be also in a separate package like googledevices.

We can either fork it, get us added as maintainers there or create something new from the scratch by just extracting some code from the component.

**Reply by ArnyminerZ** [reply]
I agree with @leikoilja because if there's a small fix to be made on a package, it's harder to get it fixed on the others, if everything is on the same package it's easier. But @KapJI also has a point on specialization, since having different packages doesn't mean that it's harder to find errors, it would be even easier, because if we know exactly what each package does, it can be easier to make fixes, and add new features.
Also, on the `googledevices` package i find it to be a bit too large for our needs, we want something simpler, just requests and parsing, so it can be a good option to take some code from them, but not the whole project, since I think it has a different focus.
Overall, I'd personally create a new package, because of specialization; even though it's harder to create updates, I think this approach should be more maintenable. But I'm open for both options, so whatever you want, will be good for me.

**Reply by leikoilja** [reply]
Okay, gents, i have just set up a [repo](https://github.com/leikoilja/google-home-api) where we can start tweaking things together. I have intentionally left the main directory `google-home-api` empty so we can start with a whole new structure including typing (from the ground up in every commit) and unit testing. This is going to be BIG! 🤩 
I have also made all the collaborators the same as this repo. 

Any thoughts? I have also enabled `discussions` in that repo where we can discuss things related to that particular project :) 🚀

**Comment by DurgNomis-drol** [comment]
Hey guys!! Glad to see you all are still working on it! 

For me the integration have not worked for a while, i am having the problem described here #202. It is very severe for me and makes my Google Home unstable and the same for my Home Assistant installation. After disabling it works again. 😭 

I will just say that i won't be actively helping anymore, but will monitor to see if a solution is found or maybe i am lucky enough that google will fix there software at some point 🤞 

I will also not continue maintaining the card i have created. But if anyone of you are interested in doing so, you are more then welcome to be a collaborator on it or fork it. Or else I will deprecated it very soon. 😢

**Comment by v1027m** [comment]
I'm sorry if this info is out there, but I can't seem to find it.


I would like to know if the NodeRed flow linked below the Readme can work for iPhones using the IOS App. It states Android, but nowhere does it mention IOS or not.

Thank you.

**Reply by leikoilja** [reply]
@mendevi, you need to contact @mattieha who has implemented the NodeRed flow. You can try reaching him out straight or mention on the PR he created for example [here](https://github.com/leikoilja/ha-google-home/pull/203)

**Comment by stboch** [comment]
@leikoilja  FYI this shouldn't be an issue since this is already the version you are using but looks like the base version of the google-api-python-client is finally getting updated... 
https://github.com/home-assistant/core/pull/67298

**Reply by leikoilja** [reply]
Awesome! Thanks for the heads up, @stboch. We have just specified that requirement in `manifest.json` (https://github.com/leikoilja/ha-google-home/pull/472), however we can probably drop it in the next HA release :)
